mypy/checker.py

```diff
@@ -834,54 +831,46 @@ def check_reverse_op_method(self, defn: FuncItem, typ: CallableType,
                                    [None, None],
                                    AnyType(TypeOfAny.special_form),
                                    self.named_type('builtins.function'))
-        if not is_subtype(typ, method_type):
-            self.msg.invalid_signature(typ, context)
+        if not is_subtype(reverse_type, method_type):
+            self.msg.invalid_signature(reverse_type, context)
             return
 
-        if method in ('__eq__', '__ne__'):
+        if reverse_name in ('__eq__', '__ne__'):
             # These are defined for all objects => can't cause trouble.
             return
 
         # With 'Any' or 'object' return type we are happy, since any possible
         # return value is valid.
-        ret_type = typ.ret_type
+        ret_type = reverse_type.ret_type
         if isinstance(ret_type, AnyType):
             return
         if isinstance(ret_type, Instance):
             if ret_type.type.fullname() == 'builtins.object':
                 return
 
-        if len(typ.arg_types) == 2:
-            # TODO check self argument kind
+        assert len(reverse_type.arg_types) == 2
 
-            # Check for the issue described above.
-            arg_type = typ.arg_types[1]
-            other_method = nodes.normal_from_reverse_op[method]
-            if isinstance(arg_type, Instance):
-                if not arg_type.type.has_readable_member(other_method):
-                    return
-            elif isinstance(arg_type, AnyType):
-                return
-            elif isinstance(arg_type, UnionType):
-                if not arg_type.has_readable_member(other_method):
-                    return
-            else:
-                return
+        forward_name = nodes.normal_from_reverse_op[reverse_name]
+        forward_base = reverse_type.arg_types[1]
+        if isinstance(forward_base, (FunctionLike, TupleType, TypedDictType)):
```

Comment:
Could you add a test for one or all of these cases, which look new?

Reply:
Added a test for TupleType. I'm not sure how to do it for the other two, if at all possible.

Reply:
@msullivan what do you think?